### PR TITLE
feat: Add structs for `GraphPath`, `PathTree`, and `FetchDependencyGraph`

### PR DESCRIPTION
**Pull request by sachindshinde**
This PR adds data structures and doc strings for:
- `GraphPath`
  - Represents a path within a query graph.
- `PathTree` 
  - Represents a collection of paths, condensed into a tree.
- `FetchDependencyGraph`
  - Represents a graph of fetches/fetch groups, where there's an edge from one fetch to another if the tail depends on the head.
- `QueryPlanningTraversal`
  - Represents the state and configuration that's maintained while generating a query plan.
 
This PR also adds various other dependent and consequent data structures.

Part of #30.

### CODE REVIEW

File: src/query_graph/graph_path.rs
Comment (SimonSapin) [inline src/query_graph/graph_path.rs:64]: Rather than have duplicated data that could accidentally get out of sync, would a method provide the same convenience? This same info is in `self.edges.first()`, right?

(Same for `tail` and `edge_to_tail`)

File: src/query_graph/graph_path.rs
Comment (SimonSapin) [inline src/query_graph/graph_path.rs:196]: Can we ever have more than two of these? Or could the Vec be replaced with `include: Option<BooleanOrVariable>, skip: Option<BooleanOrVariable>`?

File: src/query_graph/graph_path.rs
Comment (SimonSapin) [inline src/query_graph/graph_path.rs:201]: Is the intent to replace this before merging?

**Review by SimonSapin — APPROVED** [review]

File: src/query_graph/graph_path.rs
Comment (sachindshinde) [reply src/query_graph/graph_path.rs]: I think the `head`/`tail` actually point to the first/last non-null edge in the list (and if no such edges exist, `head`/`tail` point to the same node). I've updated the doc strings for them to be more clear at least. I think we can eliminate `edge_to_tail` (in the JS codebase it was eliminating two layers of indirection while here in Rust it's only eliminating one). Similarly, I'll also remove `SubgraphEnteringEdgeInfo.edge`.

File: src/query_graph/graph_path.rs
Comment (sachindshinde) [reply src/query_graph/graph_path.rs:143]: Yeah, there can be a variable number (the example just happens to have two).

File: src/query_graph/graph_path.rs
Comment (sachindshinde) [reply src/query_graph/graph_path.rs:148]: Nah, a separate PR I file will fill in the implementation (this PR is mainly to unblock work that overlaps a bit).

**Comment by sachindshinde** [comment]
Note that I ended up adding in a struct definition for `QueryPlanningTraversal` to unblock some work this week on #77. (I also tweaked operations to use `LinkedHashMap` to ensure removals don't change order.)

**Pull request merged**
